storage: error injection for disk mutations

VladLazar opened the pull request:
This PR introduces error injection capabilities to the storage layer.
The RFC can be found [here](https://docs.google.com/document/d/1Cvv-UyXNq51Rj3IjgcpARg9PHWnHHQR1AFscmlpPQdA/edit#heading=h.fxgc4ycrks86); I've not strayed too far from it, so it can be used as a rough guide.

### General approach description
A file is provided on disk via a node configuration which contains the failure injector configuration in JSON format.
This file is read from disk at start-up, then it's parsed and the required metadata is propagated through the storage
layer. This metadata is then used to construct an `ss::file` wrapper that performs failure injections according to the
config.

### Failure Injection Configuration
The JSON schema for the failure config is defined in `storage/file_sanitizer_types.h`. It allows for injecting failure on `write`, `falloc`, `truncate` and close operations. Each ntp can be configured differently and `write` operations also support batch type granularity.

Two types of failures are supported: `EIO` failures and configurable delays.

### Node Configurations
Failure injection runs independently on each node. Two new node configurations have been added to drive it:
* `storage_failure_injection_config_path` is an optional config that
  points to an on-disk path which contains the failure injection
  configuration. When this property is set, Redpanda will attempt to
  read the config and propagate it throughout the storage layer.
  Failures to read or parse this file are not critical, so are simply
  logged and ignored.
* `use_storage_failure_injection` is boolean config (defaulted to false)
  that, when set to true, enables the injection of failures and delays
  to disk mutating operations. A failure injection configuration
  specified throught the property above must also be present for
  failures to be injected.

### Potential Questions:
* Why JSON and not Serde?
I'm planning on using this with ducktape tests, so using a human readable format is easiest. If I went for serde,
I would have had to write a tool that translates from a human readable format to serde and include it in the ducktape
env.
* Is failure injection deterministic?
As much as possible, yes. A seed must be included in the failure injection configuration.
The failure configuration for each ntp uses combined hash obtained from the original seed and the ntp name hash.
Hence, given the same sequence of writes and the same seed, failures will be injected in the same manner. Achieving reproducible sequence of writes is difficult in practice, but we can take certain steps on the client side (e.g. disable batching).


Fixes https://github.com/redpanda-data/redpanda/issues/7714

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [X] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v23.1.x
- [ ] v22.3.x
- [ ] v22.2.x

## Release Notes
### Features
* Introduce file-system error injection capabilities for use in testing.

dotnwat commented:
/ci-repeat 1

jcsp commented on src/v/config/node_config.cc:
```diff
@@ -155,6 +155,20 @@ node_config::node_config() noexcept
       "Enables log messages for allocations greater than the given size.",
       {.visibility = visibility::tunable},
       std::nullopt)
+  , use_storage_failure_injection(
```
Prefer to do noun-verb style names so that both properties start with storage_failure_injection and would get alphabetically sorted together.

Something like "storage_failure_injection_enable"?

I was also wondering about the use case of a separate bool property rather than just relying on null-ness of the path property?

VladLazar replied:
> Something like "storage_failure_injection_enable"?

Sure, let's go for that.

> I was also wondering about the use case of a separate bool property rather than just relying on null-ness of the path property?

Decoupling the two makes things more flexible, since the failure config is only parsed at start-up. For instance, it allows for starting with failure injection disabled (warm-up period) and enabling it later on. I plan to add a commit to toggle failure injection via admin api. That will work by setting the boolean node config.

VladLazar replied:
Updated the config name to  `storage_failure_injection_enabled` and added a debug route to toggle failure injection at runtime.

VladLazar commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/da6f1e7b7cd902d946b05417e690e4bf565047ba..316de219e646a14e2180d2bba87286678e0c9361):
* https://github.com/redpanda-data/redpanda/compare/da6f1e7b7cd902d946b05417e690e4bf565047ba..316de219e646a14e2180d2bba87286678e0c9361

VladLazar commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/316de219e646a14e2180d2bba87286678e0c9361..8ad9d627c2f62fd924a2a1ecebf83b84f695adeb):
* Rebased and fixed conflicts

jcsp commented on src/v/storage/file_sanitizer.h:
```diff
@@ -99,9 +103,14 @@ class file_io_sanitizer final : public ss::file_impl {
             .then([this, pos, buffer, len, &pc]() {
+                return get_file_impl(_file)
+                  ->write_dma(pos, buffer, len, pc)
+                  .finally([this]() {
```
There is some cost to having the `finally(` part+lambda if we are not doing failure injection -- can we somehow make this an `unlikely()` conditional and do something like:

`auto f = write_dma()
if unlikely(failure injection) {
return f.finally..
} else {
return 
}
```
?

VladLazar replied:
Good idea. Yes, we can. I'll update the PR.

VladLazar commented on src/v/config/node_config.cc:
```diff
@@ -155,6 +155,20 @@ node_config::node_config() noexcept
+  , storage_failure_injection_enabled(
+      *this,
+      "storage_failure_injection_enabled",
+      "If true, inject low level storage failures on the write path. **Not** "
```
Failure injection is currently only done on mutating operations. We can always update the description if that changes in the future.

VladLazar commented on src/v/storage/segment_appender.cc:
```diff
@@ -111,6 +111,8 @@ segment_appender::segment_appender(segment_appender&& o) noexcept
 }
 
 ss::future<> segment_appender::append(const model::record_batch& batch) {
+    _batch_types_to_write |= 1U << static_cast<uint8_t>(batch.header().type);
```
Good idea. Let's add a static-assert somewhere.

VladLazar replied:
I added a static asssert next to the definition of `_batch_types_to_write` in `segment_appender.h`. This amazing language doesn't have a way of getting the max value of an enum, so I had to hack a `MAX` batch type in.

VladLazar commented on src/v/storage/segment_utils.cc:
```diff
@@ -166,18 +166,42 @@ ss::future<segment_appender_ptr> make_segment_appender(
+          // so we create it inline and give it the pointer after
+          // the appender is created.
+          ss::shared_ptr<file_io_sanitizer> sanitized_writer{nullptr};
+          if (ntp_sanitizer_config) {
```
> IIUC this passes the same config to every segment appender, which I believe correspond 1:1 with each segment.

Correct.

> Does that mean no matter how many segments are written, we will use the same seed on each segment within a given NTP?

Very good point! You're right. It's a tad clunky as each segment of an NTP will get the same sequence of failures.

> perhaps we should maintain some log/partition-wide seed generator that gets shared across all segments

I'd really like to avoid keeping any global state here. I'm thinking of hashing the ntp-seed with the filename, which would give us a unique seed for each segment appender.

VladLazar replied:
I ended up hashing the global seed with the file path. This should be deterministic, provided segment size jitter is disabled.

andrwng replied:
I think this is reasonable.

Maybe we should consider disabling jitter if error injection is on to make the usage of this deterministic by default. It seems like an easy piece to forget, and a test author may not think they want deterministic error injection until after it fails and they go on debugging.

VladLazar replied:
Generally, I don't like it when setting one config results in sneakily setting other stuff. We can achieve the same with some ducktape utils. I don't feel too strongly about this, so if you're convinced I'm happy to do it.

VladLazar commented on src/v/redpanda/application.cc:
```diff
@@ -1581,11 +1597,16 @@ void application::wire_up_bootstrap_services() {
+      },
+      [this, sanitizer_config]() {
+          auto log_cfg = manager_config_from_global_config(
+            _scheduling_groups, sanitizer_config);
```
There's a reason why I didn't move here. `kvstore_config_from_global_config` and `manager_config_from_global_config` both need a copy of sanitizer config. In theory, I could get away with one copy and move for the second one, but the order of evaluation for function arguments is [unspecified](https://en.cppreference.com/w/cpp/language/eval_order).

andrwng replied:
Hmm aren't we guaranteed lifecycle by virtue of the lambda captures making copies? Or are these lambdas somehow interacting with each others' captures?

Alternatively, if the lambda captures were like `[c = sanitizer_config]` could we `std::move(c)` in the lambda bodies?

Anyway, it's just a nit if you don't feel like making the change

VladLazar replied:
Done.

VladLazar commented on src/v/storage/file_sanitizer.h:
```diff
@@ -220,6 +256,74 @@ class file_io_sanitizer final : public ss::file_impl {
           [pending = std::move(pending)] {});
     }
 
+    ss::future<> maybe_inject_failure(failable_op_type op_type) {
```
Right. Scheduling points between the two places that generate random numbers should be avoided to keep it deterministic. It also feels more natural to implement such chaining via futures. I can add a comment.

VladLazar replied:
Added a comment to the impl.

VladLazar commented on src/v/storage/file_sanitizer.h:
```diff
@@ -220,6 +256,74 @@ class file_io_sanitizer final : public ss::file_impl {
+
+        if (
+          fail_cfg->delay_probability && fail_cfg->max_delay_ms
+          && fail_cfg->max_delay_ms) {
```
Fixed.

andrwng reviewed (commented):
Overall LGTM, just one more suggestion about determinism

Also I imagine there will be some follow-up PRs to plumb and exercise the new injection in ducktape?

VladLazar commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/a0ef5c6afafdf61d948ad315959e740e67d3ca03..88dc18a64dcd71f2cff294c61b8e7406172845e7):
* Use real numbers for failure injection probabilities
* Move config in `application.cc`

VladLazar commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/88dc18a64dcd71f2cff294c61b8e7406172845e7..795fbb6c182f408d3af882ef3c4e7463bc8eb48d):
* Rebased on dev to solve conflicts
* Refined failure injection logging: added a new `finject` logger and refined the sanitizer violation logs

VladLazar commented:
Chnages in [force-push](https://github.com/redpanda-data/redpanda/compare/795fbb6c182f408d3af882ef3c4e7463bc8eb48d..11b6130066a0223ff6880d98f40e7fcbcad404ea):
* Fixed a UB deref bug
* Threaded the sanitizer config into the `compaction_config` (missed that previously)

VladLazar commented:
Failure are:
* https://github.com/redpanda-data/redpanda/issues/10497 - new, but unrelated
* https://github.com/redpanda-data/redpanda/issues/9284

emaxerrno commented:
woah @VladLazar  - this changeset is amazing. well done! 

cc: @weimeilin79  & @dhruv-gupta  - we should write about this to the world. 

@VladLazar we should have you present this work on a conf.

dhruv-gupta commented:
Does this mean this is shipping in 23.2 or is this already available now?

VladLazar commented:
> Does this mean this is shipping in 23.2 or is this already available now?

This change is intended for internal testing purposes and not for general users. In that view, it's available now.
